subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformIsolationTest.groovy

```diff
@@ -0,0 +1,194 @@
+                    println "variants: " + first.collect { it.variant.attributes }
+                }
+            }
+            
+            task increment {
+                doFirst {
+                    // Just to show that incrementing the counter doesn't matter.
+                    counter.increment()
+                }
+            }
+
+            task resolveSecond(type: Copy) {
+                def second = configurations.compile.incoming.artifactView {
+                    attributes { it.attribute(artifactType, 'secondCount') }
+                }.artifacts
+                from second.artifactFiles
+                into "\${buildDir}/libs2"
+                
+                doLast {
+                    println "files: " + second.collect { it.file.name }
+                    println "ids: " + second.collect { it.id }
+                    println "components: " + second.collect { it.id.componentIdentifier }
+                    println "variants: " + second.collect { it.variant.attributes }
+                }
+            }
+
+            task resolveThird(type: Copy) {
+                def third = configurations.compile.incoming.artifactView {
+                    attributes { it.attribute(artifactType, 'thirdCount') }
+                }.artifacts
+                from third.artifactFiles
+                into "\${buildDir}/libs3"
+                
+                doLast {
+                    println "files: " + third.collect { it.file.name }
+                    println "ids: " + third.collect { it.id }
+                    println "components: " + third.collect { it.id.componentIdentifier }
+                    println "variants: " + third.collect { it.variant.attributes }
+                }
+            }
+            
+            task resolve dependsOn 'resolveFirst', 'increment', 'resolveSecond', 'resolveThird'
+        """
+
+        when:
+        run 'resolve'
+
+        then:
+        outputContains("variants: [{artifactType=firstCount}, {artifactType=firstCount}]")
+        file("build/libs1").assertHasDescendants("test-1.3.jar.txt", "test2-2.3.jar.txt")
+        file("build/libs1/test-1.3.jar.txt").readLines() == ["0", "1", "2", "3", "4"]
+        file("build/libs1/test2-2.3.jar.txt").readLines() == ["0", "1", "2", "3", "4"]
+
+        and:
+        outputContains("variants: [{artifactType=secondCount}, {artifactType=secondCount}]")
+        file("build/libs2").assertHasDescendants("test-1.3.jar.txt", "test2-2.3.jar.txt")
+        file("build/libs2/test-1.3.jar.txt").readLines() == ["1", "2", "3", "4", "5"]
+        file("build/libs2/test2-2.3.jar.txt").readLines() == ["1", "2", "3", "4", "5"]
+
+        and:
+        outputContains("variants: [{artifactType=thirdCount}, {artifactType=thirdCount}]")
+        file("build/libs3").assertHasDescendants("test-1.3.jar.txt", "test2-2.3.jar.txt")
+        file("build/libs3/test-1.3.jar.txt").readLines() == ["2", "3", "4", "5", "6"]
```

Comment:
After discussion with @oehme, it is clear that we aren't 100% aligned on what we're trying to build.

@adammurdoch told me this morning in a slack conversation:
> we want to verify that
> 1. none of the transforms see changes made by the build logic after registration, and
> 2. none of the transforms see changes made by the other transforms

I think we are all agreed on part 2. But, @oehme makes a good case for NOT wanting part 1.

Essentially, snapshotting at transform eagerly at registration time means that the user can no longer configure the parameters to the transform after it's registered. But, that's exactly what they will sometimes want to do. For example, a plugin might register a transform that depends on an Android build flavor, and then later in the configuration, in the user's build script, they may configure the build flavor to use some custom dexer. We want to notice that change and perform the correct transform "in isolation()" only just before we execute the transform.

@adammurdoch, please respond whether or not you agree with @oehme's recommendation to defer both snapshotting AND isolation until just before applying the transform.